pkg/tracing/google_cloud/google_cloud.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package google_cloud
+
+import (
+	"context"
+	"os"
+
+	cloudtrace "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/version"
+	"github.com/thanos-io/thanos/pkg/tracing/migration"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+	"gopkg.in/yaml.v2"
+)
+
+// Config - YAML configuration.
+type Config struct {
+	ServiceName  string `yaml:"service_name"`
+	ProjectId    string `yaml:"project_id"`
+	SampleFactor uint64 `yaml:"sample_factor"`
+}
+
+// NewTracerProvider create tracer provider from YAML.
+func NewTracerProvider(ctx context.Context, logger log.Logger, conf []byte) (*tracesdk.TracerProvider, error) {
+	config := Config{}
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, err
+	}
+
+	exporter, err := cloudtrace.NewExporter(
+		cloudtrace.WithContext(ctx),
+		cloudtrace.WithProjectID(config.ProjectId),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTracerProvider(ctx, logger, tracesdk.NewBatchSpanProcessor(exporter),
+		config.SampleFactor, config.ServiceName), nil
+}
+
+func newTracerProvider(ctx context.Context, logger log.Logger, processor tracesdk.SpanProcessor, sampleFactor uint64, serviceName string) *tracesdk.TracerProvider {
+	fraction := 1 / float64(sampleFactor)
+	if sampleFactor == 0 {
+		fraction = 0
+	}
+
+	resource, err := resource.New(ctx, resource.WithAttributes(collectAttributes(serviceName)...))
+	if err != nil {
+		level.Warn(logger).Log("msg", "detecting resources for tracing provider failed", "err", err)
```

Comment:
If there is an error, should we return it or do we need to deal with it in line 67 as well?

Reply:
If there's an error, we should still receive valid (although possibly empty) `resource`. I think best we can do is log warning and move on (or we can possibly change level to `error`). I'll add a comment as well.